File: cses/04_graph_algorithms/1160_planets-queries-ii/solution.cpp
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, q;
    std::cin >> n >> q;

    std::vector<int> next(n + 1);
    for (int v = 1; v <= n; ++v) {
        std::cin >> next[v];
    }

    // Bước 1: tìm mọi cycle bằng cách duyệt functional graph theo kiểu iterative.
    // state: 0 = chưa thăm, 1 = đang trên đường đi hiện tại, 2 = đã xong.
    std::vector<int> state(n + 1, 0);
    std::vector<int> path_index(n + 1, 0);
    std::vector<char> on_cycle(n + 1, false);

    for (int start = 1; start <= n; ++start) {
        if (state[start] != 0) {
            continue;
        }

        std::vector<int> path;
        int v = start;
        while (state[v] == 0) {
            state[v] = 1;
            path_index[v] = static_cast<int>(path.size());
            path.push_back(v);
            v = next[v];
        }

        // Gặp lại một đỉnh đang trên đường đi hiện tại -> phát hiện cycle mới.
        if (state[v] == 1) {
            for (int i = path_index[v]; i < static_cast<int>(path.size()); ++i) {
                on_cycle[path[i]] = true;
            }
        }
        for (int node : path) {
            state[node] = 2;
        }
    }

    // Bước 2: đánh số các cycle và ghi lại vị trí của từng đỉnh theo thứ tự đi vòng.
    std::vector<int> cycle_id(n + 1, -1);
    std::vector<int> cycle_position(n + 1, 0);
    std::vector<int> cycle_length;

    for (int v = 1; v <= n; ++v) {
        if (!on_cycle[v] || cycle_id[v] != -1) {
            continue;
        }

        const int id = static_cast<int>(cycle_length.size());
        int current = v;
        int position = 0;
        do {
            cycle_id[current] = id;
            cycle_position[current] = position++;
            current = next[current];
        } while (current != v);
        cycle_length.push_back(position);
    }

    // Bước 3: bỏ các cạnh cycle -> rừng cây với gốc là các đỉnh cycle.
    // Xây danh sách con theo cạnh ngược: u là con của next[u] nếu u không nằm trên cycle.
    std::vector<std::vector<int>> children(n + 1);
    for (int v = 1; v <= n; ++v) {
        if (!on_cycle[v]) {
            children[next[v]].push_back(v);
        }
    }

    // DFS iterative từ mỗi đỉnh cycle để tính depth, root và Euler tour (tin/tout).
    std::vector<int> depth(n + 1, 0);
    std::vector<int> root(n + 1, 0);
    std::vector<int> tin(n + 1, 0);
    std::vector<int> tout(n + 1, 0);
    std::vector<int> child_index(n + 1, 0);
    int timer = 0;

    for (int cycle_vertex = 1; cycle_vertex <= n; ++cycle_vertex) {
        if (!on_cycle[cycle_vertex]) {
            continue;
        }

        root[cycle_vertex] = cycle_vertex;
        tin[cycle_vertex] = timer++;
        std::vector<int> stack{cycle_vertex};

        while (!stack.empty()) {
            const int node = stack.back();
            if (child_index[node] < static_cast<int>(children[node].size())) {
                const int child = children[node][child_index[node]++];
                depth[child] = depth[node] + 1;
                root[child] = cycle_vertex;
                tin[child] = timer++;
                stack.push_back(child);
            } else {
                tout[node] = timer - 1;
                stack.pop_back();
            }
        }
    }

    // Bước 4: trả lời từng truy vấn (a, b).
    while (q-- > 0) {
        int a, b;
        std::cin >> a >> b;

        if (on_cycle[b]) {
            // b nằm trên cycle: chỉ tới được nếu a và b cùng component.
            const int entry = root[a];
            if (cycle_id[entry] != cycle_id[b]) {
                std::cout << -1 << '\n';
                continue;
            }

            // depth[a] bước tới đỉnh vào cycle, cộng khoảng cách đi vòng tới b.
            const int length = cycle_length[cycle_id[b]];
            const int around =
                (cycle_position[b] - cycle_position[entry] + length) % length;
            std::cout << depth[a] + around << '\n';
        } else if (tin[b] <= tin[a] && tin[a] <= tout[b]) {
            // b nằm trong cây và là tổ tiên của a: đáp án là hiệu độ sâu.
            std::cout << depth[a] - depth[b] << '\n';
        } else {
            std::cout << -1 << '\n';
        }
    }

    return 0;
}

File: cses/04_graph_algorithms/1160_planets-queries-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1
    q = int(data[pos]); pos += 1

    t = [0] * (n + 1)
    for i in range(1, n + 1):
        t[i] = int(data[pos]); pos += 1

    # ---- 1) Tìm các đỉnh nằm trên cycle (iterative) ----
    state = [0] * (n + 1)      # 0 chưa thăm, 1 đang trên đường đi, 2 đã xong
    order = [0] * (n + 1)      # chỉ số của đỉnh trong đường đi hiện tại
    on_cycle = bytearray(n + 1)
    for start in range(1, n + 1):
        if state[start] != 0:
            continue
        path = []
        v = start
        while state[v] == 0:
            state[v] = 1
            order[v] = len(path)
            path.append(v)
            v = t[v]
        if state[v] == 1:  # gặp lại đỉnh đang trên đường đi -> cycle mới
            si = order[v]
            for j in range(si, len(path)):
                on_cycle[path[j]] = 1
        for node in path:
            state[node] = 2

    # ---- 2) Gán cycle_id, pos_in_cycle, độ dài cycle ----
    cycle_id = [-1] * (n + 1)
    pos_in_cycle = [0] * (n + 1)
    cyclen = []
    cid = 0
    for v in range(1, n + 1):
        if on_cycle[v] and cycle_id[v] == -1:
            cur = v
            p = 0
            while cycle_id[cur] == -1:
                cycle_id[cur] = cid
                pos_in_cycle[cur] = p
                p += 1
                cur = t[cur]
            cyclen.append(p)
            cid += 1

    # ---- 3) Rừng cây (cạnh ngược, bỏ cạnh cycle) ----
    children = [[] for _ in range(n + 1)]
    for u in range(1, n + 1):
        if not on_cycle[u]:
            children[t[u]].append(u)

    depth = [0] * (n + 1)
    root = [0] * (n + 1)
    in_t = [0] * (n + 1)
    out_t = [0] * (n + 1)
    it = [0] * (n + 1)   # con trỏ con cho DFS iterative

    timer = 0
    for c in range(1, n + 1):
        if not on_cycle[c]:
            continue
        root[c] = c
        depth[c] = 0
        in_t[c] = timer
        timer += 1
        stack = [c]
        while stack:
            node = stack[-1]
            i = it[node]
            chl = children[node]
            if i < len(chl):
                it[node] = i + 1
                ch = chl[i]
                depth[ch] = depth[node] + 1
                root[ch] = c
                in_t[ch] = timer
                timer += 1
                stack.append(ch)
            else:
                out_t[node] = timer - 1
                stack.pop()

    # ---- 4) Trả lời truy vấn ----
    out = []
    ap = out.append
    for _ in range(q):
        a = int(data[pos]); b = int(data[pos + 1]); pos += 2
        if on_cycle[b]:
            ra = root[a]
            cb = cycle_id[b]
            if cycle_id[ra] == cb:
                L = cyclen[cb]
                ap(depth[a] + (pos_in_cycle[b] - pos_in_cycle[ra]) % L)
            else:
                ap(-1)
        else:
            if in_t[b] <= in_t[a] <= out_t[b]:
                ap(depth[a] - depth[b])
            else:
                ap(-1)

    sys.stdout.write('\n'.join(map(str, out)))
    if out:
        sys.stdout.write('\n')


main()
